mempool/clist_mempool.go

```diff
@@ -643,9 +644,33 @@ type mempoolTx struct {
 	gasWanted int64    // amount of gas this tx states it will require
 	tx        types.Tx //
 
-	// ids of peers who've sent us this tx (as a map for quick lookups).
-	// senders: PeerID -> bool
-	senders sync.Map
+	// mu guards following fields
+	mu sync.Mutex
+	// senders is a bitset ids of peers who've sent us this tx
+	// TODO: make libs/bits.BitArray has ready to use zero value and use it instead.
+	senders big.Int
+}
+
+func (memTx *mempoolTx) LoadOrStoreSender(senderID uint16) bool {
+	memTx.mu.Lock()
+	defer memTx.mu.Unlock()
+	if memTx.senders.Bit(int(senderID)) != 0 {
+		return true
+	}
+	memTx.senders.SetBit(&memTx.senders, int(senderID), 1)
+	return false
+}
+
+func (memTx *mempoolTx) AddSender(senderID uint16) {
+	memTx.mu.Lock()
+	defer memTx.mu.Unlock()
+	memTx.senders.SetBit(&memTx.senders, int(senderID), 1)
+}
+
+func (memTx *mempoolTx) HasSender(senderID uint16) bool {
+	memTx.mu.Lock()
+	defer memTx.mu.Unlock()
```

Comment:
Please a new line after every defer unlock, that helps delineate code logic sections thus making it more readable.